Applications/Calculator/CalculatorWidget.cpp

```diff
@@ -0,0 +1,204 @@
+    button_backspace.set_relative_rect(55, 40, 65, 28);
+
+    auto& button_decimal_point = *new GButton(this);
+    button_decimal_point.move_to(133, 177);
+    button_decimal_point.set_foreground_color(Color::NamedColor::Blue);
+    button_decimal_point.set_text(".");
+    button_decimal_point.on_click = [this](GButton&) {
+        m_keypad.type_decimal_point();
+        update_display();
+    };
+    add_button(button_decimal_point);
+
+    auto& button_toggle_sign = *new GButton(this);
+    button_toggle_sign.move_to(94, 177);
+    button_toggle_sign.set_foreground_color(Color::NamedColor::Blue);
+    button_toggle_sign.set_text("+/-");
+    add_button(button_toggle_sign, Calculator::Operation::ToggleSign);
+
+    auto& button_add = *new GButton(this);
+    button_add.move_to(172, 177);
+    button_add.set_foreground_color(Color::NamedColor::Red);
+    button_add.set_text("+");
+    add_button(button_add, Calculator::Operation::Add);
+
+    auto& button_subtract = *new GButton(this);
+    button_subtract.move_to(172, 144);
+    button_subtract.set_foreground_color(Color::NamedColor::Red);
+    button_subtract.set_text("-");
+    add_button(button_subtract, Calculator::Operation::Subtract);
+
+    auto& button_multiply = *new GButton(this);
+    button_multiply.move_to(172, 111);
+    button_multiply.set_foreground_color(Color::NamedColor::Red);
+    button_multiply.set_text("*");
+    add_button(button_multiply, Calculator::Operation::Multiply);
+
+    auto& button_divide = *new GButton(this);
+    button_divide.move_to(172, 78);
+    button_divide.set_foreground_color(Color::NamedColor::Red);
+    button_divide.set_text("/");
+    add_button(button_divide, Calculator::Operation::Divide);
+
+    auto& button_sqrt = *new GButton(this);
+    button_sqrt.move_to(211, 78);
+    button_sqrt.set_foreground_color(Color::NamedColor::Blue);
+    button_sqrt.set_text("sqrt");
+    add_button(button_sqrt, Calculator::Operation::Sqrt);
+
+    auto& button_inverse = *new GButton(this);
+    button_inverse.move_to(211, 144);
+    button_inverse.set_foreground_color(Color::NamedColor::Blue);
+    button_inverse.set_text("1/x");
+    add_button(button_inverse, Calculator::Operation::Inverse);
+
+    auto& button_percent = *new GButton(this);
+    button_percent.move_to(211, 111);
+    button_percent.set_foreground_color(Color::NamedColor::Blue);
+    button_percent.set_text("%");
+    add_button(button_percent, Calculator::Operation::Percent);
+
+    auto& button_equals = *new GButton(this);
+    button_equals.move_to(211, 177);
+    button_equals.set_foreground_color(Color::NamedColor::Red);
+    button_equals.set_text("=");
+    button_equals.on_click = [this](GButton&) {
+        double argument = m_keypad.value();
+        double res = m_calculator.finish_operation(argument);
+        m_keypad.set_value(res);
+        update_display();
+    };
+    add_button(button_equals);
+}
+
+CalculatorWidget::~CalculatorWidget()
+{
+    delete m_entry;
```

Comment:
No need for any of the `delete` calls here. These are all `CObject`s, and they will be destroyed by their parent.
See `CObject::~CObject` for the destruction.

Reply:
Ouch, haven't thought of that.

This is kind of weird though, I'm writing `new`, so it should also be me who `delete`s it. If I'm passing ownership to the parent, I'd expect there to be some smart pointer type that communicates that it's happening and ensures it actually happens as it should. I suppose the problem is that we're not passing the child to `parent->add_child()`, we're passing the parent into the child's constructor, so there's no way to wrap the child into a smart pointer by that time.

Reply:
Yeah, this is based on the Qt framework’s object model so anyone who has been using that before will recognize the pattern.
It is weird and doesn’t feel right though. That’s why I intend to replace it with reference counting, it’s just a big project that I keep postponing..